pkg/encode/encoder.go

```diff
@@ -0,0 +1,451 @@
+/*
+ * Copyright 2023 CloudWeGo Authors
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package encode
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+	"reflect"
+	"strings"
+
+	commons "github.com/kitex-contrib/codec-hessian2/pkg/common"
+)
+
+func NewHessian2Output(w io.Writer) *Encoder {
+	return &Encoder{
+		writer: w,
+		buffer: bytes.NewBuffer(nil),
+	}
+}
+
+// Class表示Java对象的类型信息
+type Class struct {
+	Name   string   // 类名
+	Fields []*Field // 字段信息
+}
+
+// Field表示Java对象的字段信息
+type Field struct {
+	Name    string // 字段名
+	Type    *Class // 字段类型
+	Default Object // 字段默认值
+}
+
+// Object表示Java对象的值信息
+type Object interface{}
+
+type Encoder struct {
+	writer io.Writer     // 输出流
+	buffer *bytes.Buffer // buffer缓存
+
+	writeType      bool     // 是否写入类型信息
+	typeRefWritten bool     // 是否已经将类型引用写入
+	typeRefs       []*Class // 类型引用数组
+	objRefs        []Object // 对象引用数组
+}
+
+func (e *Encoder) writeObjectRef(i int) error {
+	if err := e.WriteByte('R'); err != nil {
+		return err
+	}
+	if err := binary.Write(e.buffer, binary.BigEndian, i); err != nil {
+		return err
+	}
+	return nil
+}
+func (e *Encoder) WriteObject(obj Object) error {
+	// 判断是否为nil
+	if obj == nil {
```

Comment:
Is it possbile that obj is a typed nil pointer? If so, this check is not enough, we also need to check `reflect.Valueof(obj).isNil()`